### PR TITLE
Expression having "I"(iota) be rewritten in terms of 'exp'/'log' in DifferentialExtension

### PR DESCRIPTION
**Pull request by gxyd**
Any expression containing `I` be rewritten in terms of `exp`/`log`. Also use `cancel` to remove factors from numerator and denominator of expression.

### CODE REVIEW

File: sympy/integrals/risch.py [hunk 1]
```diff
@@ -218,14 +218,14 @@ def __init__(self, f=None, x=None, handle_first='log', dummy=False, extension=No
         self.dummy = dummy
         self.reset()
         exp_new_extension, log_new_extension = True, True
-        if rewrite_complex:
+        if rewrite_complex or I in self.f.atoms():
```
Comment (jksuom) [inline sympy/integrals/risch.py]: This could be included in the definition of rewrite complex. (Set to True if `I in self.f.atoms()`.)

File: sympy/integrals/risch.py [hunk 2]
```diff
@@ -218,14 +218,14 @@ def __init__(self, f=None, x=None, handle_first='log', dummy=False, extension=No
...
             rewritables = {
                 (sin, cos, cot, tan, sinh, cosh, coth, tanh): exp,
                 (asin, acos, acot, atan): log,
             }
         # rewrite the trigonometric components
             for candidates, rule in rewritables.items():
-                self.newf = self.newf.rewrite(candidates, rule)
+                self.newf = cancel(self.newf.rewrite(candidates, rule))
```
Comment (asmeurer) [inline sympy/integrals/risch.py]: Why is this added?

File: sympy/integrals/risch.py (hunk 1 above)
Comment (asmeurer) [reply sympy/integrals/risch.py]: I would set the default value to `None`, meaning "automatic" (it does it if it detects a complex constant, and not otherwise). It can also be set to True or False manually by the user (with a caveat that False could lead to invalid nonelementary results).

File: sympy/integrals/risch.py (hunk 2 above)
Comment (gxyd) [reply sympy/integrals/risch.py]: With an example:
```python
>>> f = tan(atan(x).rewrite(log))
>>> print(f.rewrite(exp))
I*(sqrt(-I*x + 1)/sqrt(I*x + 1) - sqrt(I*x + 1)/sqrt(-I*x + 1))/(sqrt(-I*x + 1)/sqrt(I*x + 1) + sqrt(I*x + 1)/sqrt(-I*x + 1))

>>> print(cancel(f.rewrite(exp)))
x
```

File: sympy/integrals/risch.py [hunk 3]
```diff
@@ -218,14 +218,20 @@ def __init__(self, f=None, x=None, handle_first='log', dummy=False, extension=No
...
         if rewrite_complex:
             rewritables = {
                 (sin, cos, cot, tan, sinh, cosh, coth, tanh): exp,
                 (asin, acos, acot, atan): log,
             }
-        # rewrite the trigonometric components
+            # rewrite the trigonometric components
             for candidates, rule in rewritables.items():
                 self.newf = self.newf.rewrite(candidates, rule)
+            self.newf = cancel(self.newf)
```
Comment (gxyd) [inline sympy/integrals/risch.py]: I don't think we will be needing two separate (one each after `exp` and `log` rewrite) or just one after both the rewrites have been done.

Comment (gxyd) [reply sympy/integrals/risch.py]: Currently it is failing a test https://travis-ci.org/sympy/sympy/jobs/270698759 .https://github.com/sympy/sympy/blob/96074807d8d003dbd83198d82ee0d5b16390709d/sympy/integrals/tests/test_integrals.py#L897 . Though risch algorithm is not able (non-elementary) integrate it but it modifies the expression for `risch=False` as well, and that is a problem for it.

Comment (gxyd) [reply sympy/integrals/risch.py]: Can that related test be changed to `risch=False`?

Comment (asmeurer) [reply sympy/integrals/risch.py]: How does it fail the test if risch cannot integrate it?

Comment (gxyd) [reply sympy/integrals/risch.py]: The risch modifies the expression to separate it (into `exp(-x**2)*exp(I*c*x)`), that is being done because of `cancel` function.

This is similar to passing `integrate(exp(-x**2)*exp(I*c*x), x)` into current function.

Comment (asmeurer) [reply sympy/integrals/risch.py]: Does the risch rewriting affect the other integrate routines? 

Anyway, how is the test failing? If it's still the right answer in a different form that's fine.

Comment (gxyd) [reply sympy/integrals/risch.py]: It is returned as it is:
```
In [3]: integrate(exp(-x**2 + I*c*x), x)
Out[3]: 
⌠               
⎮    2          
⎮  -x   ⅈ⋅c⋅x   
⎮ ℯ   ⋅ℯ      dx
⌡               
```

Comment (gxyd) [reply sympy/integrals/risch.py]: > Does the risch rewriting affect the other integrate routines?

Yes, I think it does affect.

Comment (asmeurer) [reply sympy/integrals/risch.py]: So let's make it so that if the entire expression is nonelementary that it reuses the original expression, not the rewritten one.

File: sympy/integrals/tests/test_integrals.py [hunk 4]
```diff
@@ -1107,7 +1107,7 @@ def test_issue_2708():
     assert integrate(2*f + exp(z), (z, 2, 3)) == \
         2*integral_f - exp(2) + exp(3)
     assert integrate(exp(1.2*n*s*z*(-t + z)/t), (z, 0, x)) == \
-        1.0*NonElementaryIntegral(exp(-1.2*n*s*z)*exp(1.2*n*s*z**2/t),
+        NonElementaryIntegral(exp(-1.2*n*s*z)*exp(1.2*n*s*z**2/t),
```
Comment (gxyd) [inline sympy/integrals/tests/test_integrals.py]: I am not sure, earlier why having `1.0` here was necessary. But mathematically they are same thing.

**Comment by gxyd** [comment]
@asmeure does this look alright? It passes the tests on travis as well though the "allowed failures" remain.

**Comment by asmeurer** [comment]
It looks good to me.

**Comment by gxyd** [comment]
If it looks okay, can we please merge this PR? I will merge `master` into the `tans` branch then.

**Comment by gxyd** [comment]
@jksuom can we merge this?

File: sympy/integrals/integrals.py [hunk 5]
```diff
@@ -783,7 +783,15 @@ def _eval_integral(self, f, x, meijerg=None, risch=None, manual=None,
...
                 if i:
                     # There was a nonelementary integral. Try integrating it.
-                    return result + i.doit(risch=False)
+
+                    # if no part of the NonElementaryIntegral is integrated by
+                    # the Risch algorithm, then use the original function to
+                    # integrate, instead of re-written one
+                    if result == 0:
+                        from sympy.integrals.risch import NonElementaryIntegral
+                        return NonElementaryIntegral(f, x).doit(risch=False)
```
Comment (jksuom) [inline sympy/integrals/integrals.py]: How does `NonElementaryIntegral.doit` differ from `Integral.doit`? What if `risch=False` is not added?

Comment (gxyd) [reply sympy/integrals/integrals.py]: > What if risch=False is not added?

`risch=True` has already been tried above. Resulting in a conclusion that the integral as a whole is `NonElementary`. Also in this `if` condition `result == 0` implies, that no part of the integral was computed by Risch algorithm.

Comment (gxyd) [reply sympy/integrals/integrals.py]: > How does NonElementaryIntegral.doit differ from Integral.doit?

`NonElementaryIntegral` doesn't differ in any way from using `Integral`, but I wanted to make it explicitly clear that integral has been proved to be non-elementary here. Otherwise even in working there would be no difference.

Comment (gxyd) [reply sympy/integrals/integrals.py]: It would keep on trying the `risch=True` first. Even below, you can see that it adds `risch=False` to since we are in a block of code, where we know that Risch algorithm, wouldn't be able 'fully' integrate the given function.

This piece of doesn't do anything fancy. Just checks whether there has been any integral computation done by Risch algorithm or not. If not then the integral with `risch=False` be tried with the original `f` (function) independent of risch from rewrites of `f`.

Comment (jksuom) [reply sympy/integrals/integrals.py]: Ok, thanks. I was wondering since Risch integration is the only area where the difference between elementary and nonelementary integrals is significant. So I thought that having a plain `Integral` with the original function and `risch=False` could have been enough. But maybe `NonElementaryIntegral` would be useful if the integral will remain unevaluated.

**Comment by jksuom** [comment]
I think it is ok to merge.

**Comment by Upabjojr** [comment]
Hi, did this PR break some integrals by chance?
```
In [5]: integrate(exp(x)**x, x)
Out[5]: 
⌠         
⎮     x   
⎮ ⎛ x⎞    
⎮ ⎝ℯ ⎠  dx
⌡     
```

**Comment by gxyd** [comment]
I see, that change made by Risch algorithm to the expression is 'used' (intentionally/unintentionally) the non-risch algorithm part and is able to solve it.

Below are the results from previous master
```
In [9]: integrate(exp(x)**x, x, risch=True)
Out[9]: Integral(exp(x**2), x)   # the risch algorithm modifies the expression

In [10]: integrate(exp(x)**x, x, risch=False)
Out[10]: Integral(exp(x)**x, x)  # the non-risch part is not able to solve it on its own

In [11]: integrate(exp(x)**x, x)
Out[11]: sqrt(pi)*erfi(x)/2  # the modified expression is used by the risch algorithm but doesn't solve it and is solved by the non-risch part.
```

So the question is: Does risch algorithm part of integration changes the function for good or bad when it is not able to solve it at all? Should that be further used for computation or original function be used.

**Comment by asmeurer** [comment]
I think any modifications that risch makes should be kept internal to the algorithm. They might help or hurt other algorithms, but they shouldn't be considered dependent on anything other than risch itself.

**Comment by jksuom** [comment]
I don't think there is any integrator that could currently handle iterated powers `(a**b)**c` without first transforming those to `a**(b*c)`. Maybe there could be a preprocessor, common for all integrators, to handle that kind of generally beneficial transformations.

**Comment by Upabjojr** [comment]
Is that property `exp(x)**x == exp(x**2)` valid for any complex number?

```python
In [1]: (exp(x)**x).subs(x, 1 + 2*pi*I)
Out[1]: 
 1
ℯ 

In [2]: (exp(x**2)).subs(x, 1 + 2*pi*I)
Out[2]: 
 ⎛           2⎞
 ⎝(1 + 2⋅ⅈ⋅π) ⎠
ℯ                     

In [4]: (exp(x**2)).subs(x, 1 + 2*pi*I).n()
Out[4]: 1.94551938330541e-17 - 0.e-39⋅ⅈ
```

Wolfram Mathematica gives:

```
In[2]:= Integrate[(E^x)^x, x]                                                   

                               x
                      x + Log[E ]
        Sqrt[Pi] Erfi[-----------]
                           2
Out[2]= --------------------------
                         x  2
              (-x + Log[E ]) /4
           2 E
```

**Comment by asmeurer** [comment]
Interesting. Is that expression really different from `sqrt(pi)*erfi(x)/2` or does it differ by a constant. I don't know what sorts of identities are satisfied by erfi. It sounds like the integral of `exp(x)**x` now giving a nonanswer is fine if the previous answer was wrong. 

In Risch, it is assumed that `log(exp(x))` is equal to `x` (Risch has to treat expressions algebraically; to Risch, `log(x)` is just the expression whose derivative is `1/x`).  Thus `exp(x)**x` is converted to `exp(x*log(exp(x))`, which gives `exp(x**2)`. 

This is a bit worrying if it can lead to wrong results in the complex plane. My understanding has always been that it's fine because the symbolic derivatives work out the same, so any difference is a piecewise constant.

I'm curious, what does Mathematica give as the integral of `x*exp(x)**x`?

**Comment by asmeurer** [comment]
[Wolfram Alpha](https://www.wolframalpha.com/input/?i=integrate(x*exp(x)**x)&dataset=) gives 

```
(2 (E^x)^x + (Sqrt[Pi] Erfi[(x + Log[E^x])/2] (x - Log[E^x]))/E^((x - Log[E^x])^2/4))/4
```

which has the same `erfi((x + log(exp(x)))/2)` term.

**Comment by asmeurer** [comment]
I opened a new issue for this https://github.com/sympy/sympy/issues/13280

**Comment by jksuom** [comment]
The expression `exp(x)**x` can be regarded as defining a multivalued function having an infinite number of analytic branches 

    exp(x**2 + n*2*pi*I*x) (n in Z)

each one of which has an antiderivative in the complex plane that can be used to compute contour integrals.

It seems that Mathematica interprets the expression as a combination of the different branches with branch cuts along the lines `im(x) == n*pi` for odd integers `n`, more precisely, as a piecewise analytic function taking the value `exp(x**2)` in the horizontal zone `|im(x)| < pi` of width `2*pi`, the value `exp(x**2 - 2*pi*I*x)`  in the zone `|im(x) - 2*pi| < pi`, etc.. The antiderivative can then be used to compute contour integrals within a single zone. It cannot be used to compute integrals along curves that pass a branch cut.

Similar multivalued functions are currently handled by the meijerint module. Its results usually consist of pieces that are valid in some domains. `exp(x)**x` could also be dealt with in the same way. A preprocessor would transform it into more manageable forms adding conditions like `|im(x)| < pi`. That would suffice for the computation of definite integrals along the real line. A more sophisticated preprocessor could transform it into a form covering all zones by using `n*2*pi*I = log(exp(x)) - x`:

    exp(x**2 + n*2*pi*I*x) = exp((x + n*pi*I)**2)/exp((n*pi*I)**2)
                      = exp(((x + log(exp(x))/2)**2)/exp((log(exp(x)) - x)**2/4)
